### PR TITLE
Skip unsupervised teacher projections in pruned EAGLE3 targets

### PR DESCRIPTION
**Pull request by julyanghar**
For vocabulary-pruned EAGLE3 training, `compute_target_p_padded` projects and softmaxes unsupervised teacher rows even though the loss never consumes them. This PR projects supervised rows in the existing chunks, reuses pruned logits already calculated for coverage, and scatters into the original dense target lookup table. Unused rows are zero; shapes, padding, position masks and coverage semantics are preserved. The all-valid dense fallback is retained after comparison with a scatter-only alternative.

Validation: 78 relevant tests pass, including FP32/BF16 target probabilities, chunk boundaries, sparse/full/empty masks, shifted multi-round lookup, and model parameter-gradient checks. Changed files pass Ruff check/format and `git diff --check`.

The [review evidence package](https://github.com/julyanghar/TorchSpec/blob/592b589b4e79e6502f7162a3fa17084253fcdc99/experiments/pr193-review/README.md) includes real data statistics, all timing cases, NSYS diagnostics, raw trajectories, figures and reproduction scripts:

- 1,000 real training batches: supervised rows / padded rows = 69.02%; vocabulary coverage among supervised rows = 99.24%.
- Native EAGLE3 Trainer runs: 1,000 optimizer steps per arm, shared offline teacher data, same initialization/order/GPU. Two 50-step U repeats are exact. U/P are not pointwise identical: mean relative loss difference 0.591%, maximum 7.38%; final fixed-evaluation weighted loss differs by 0.00728%. Intermediate evaluation deviations and all per-step data are retained; no multi-seed or exact-trajectory equivalence claim.
- Teacher-function benchmark: sum of mean times across 12 real batch cases is 14.98% lower. This is not end-to-end training throughput. At the requested synthetic 1950/2048 geometry, the PR is 0.61% slower, also reported.
- NSYS: the `.numel()` comparison produced no CUDA API calls in 20 ranges; the existing `nonzero()` remains a synchronization point.

### CODE REVIEW

File: torchspec/models/eagle3.py [hunk 1]
```diff
@@ -477,6 +477,14 @@ def compute_target_p_padded(
         device=target_hidden_states.device,
         dtype=torch.float,
     )
+    # Keep the dense lookup layout for TTT, but project only supervised rows.
+    all_valid = valid_flat_idx.numel() == bsz * seq_len
```
Comment (Dogacel) [inline torchspec/models/eagle3.py:481]: I think `.numel()` is making a CPU call, making GPU threads wait.

**Review by Dogacel — COMMENTED** [review]
I think this makes sense, but I have several concerns.

1. Is 204/2048 a realistic sparsity? Most likely vocab pruning will prune around 5% at most right? In that case, shouldn't we test sth like 1950/2048 to test the performance? I think we should rely on training data.

2. Since valid elements is almost never empty, we can default to the index scatter path.

Also I would like to see 2 training runs for ~1k steps on EAGLE-3 to show the loss is equal as validation.

File: torchspec/models/eagle3.py (hunk 1 above)
Comment (julyanghar) [reply torchspec/models/eagle3.py:481]: I checked this with NSYS in the actual eager precompute context. `.numel()` reads tensor size metadata already available on the host; it does not retrieve tensor values from CUDA. Across 20 `numel + comparison` ranges, NSYS recorded **zero CUDA API calls**, and deliberately queued GPU work was still pending after every range. The preceding `nonzero()` and an `.item()` positive control each recorded 20 memcpy/synchronization calls. The `nonzero()` is already in the base implementation.

There is still ordinary host-call overhead; this is not a claim that the branch is free. I also benchmarked the scatter-only alternative, as summarized in my main reply. [Trace summary](https://github.com/julyanghar/TorchSpec/blob/592b589b4e79e6502f7162a3fa17084253fcdc99/experiments/pr193-review/sync-summary.json) and [probe source](https://github.com/julyanghar/TorchSpec/blob/592b589b4e79e6502f7162a3fa17084253fcdc99/experiments/pr193-review/sync_probe.py).
